From 7770796772f39ce9115e9cc84c71d57475328ecf Mon Sep 17 00:00:00 2001
From: Son Lam <son.lam.ym@rvc.renesas.com>
Date: Tue, 29 Jan 2019 09:28:02 +0700
Subject: [PATCH] omxvideodec: don't drop frame if it contains header data

handle_frame() will drop frame until a sync point is found.
It is not wrong because frame cannot be decoded properly without
a sync point (I frame).
However, if the frame with header data (SPS, PPS) is discarded
completely and never appear again, video cannot be decoded at all.

Avoid this issue by not dropping header data frame.
Note that this only change the behaviour with video which have late
sync point.
Before this change, video cannot be decoded.
After this change, video can be decoded, but some early frames
before sync points can be decoded wrongly.

Signed-off-by: Hung Tran <hung.tran.jy@renesas.com>
Signed-off-by: Son Lam <son.lam.ym@rvc.renesas.com>
---
 omx/gstomxvideodec.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 1dabfad..c5d5bf1 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -2523,7 +2523,8 @@ gst_omx_video_dec_handle_frame (GstVideoDecoder * decoder,
   GST_DEBUG_OBJECT (self, "Handling frame");
 
   if (!self->started) {
-    if (!GST_VIDEO_CODEC_FRAME_IS_SYNC_POINT (frame)) {
+    if (!GST_VIDEO_CODEC_FRAME_IS_SYNC_POINT (frame) &&
+        !GST_BUFFER_FLAG_IS_SET (frame->input_buffer, GST_BUFFER_FLAG_HEADER)) {
       gst_video_decoder_drop_frame (GST_VIDEO_DECODER (self), frame);
       return GST_FLOW_OK;
     }
-- 
2.7.4

